### ISSUE TITLE
Secondary Entry Points fail resolve path

### ISSUE THREAD

**Issue by morganpizzini** [ng-packagr/ng-packagr#1239]
## Type of Issue

```
[x] Bug Report
[ ] Feature Request
```

## Description
I build secondary entry point for holding ngrx reducers/action/services, ng-packagr fail to build the relative path to entity in main library.
- Main library path motortrack-lib
- Secondary path motortrack-lib/services
~~~
export declare const selectTripEntity: import("@ngrx/store").MemoizedSelector<object, import("../../../../../../../dist/motortrack-lib/public_api").TripContract>;
~~~
the error is ng-packagr add an extra '../' on path

#### How To Reproduce

in reducers i export entity as follow
~~~
// trips.reducer.ts
import {TripContract} form '@motortrack-lib'
// state

export interface State {
  selectedTrip: TripContract;
}

export const getSelectedTrip = (state: State) => state.selectedTrip;
~~~
then in index file i create a selector by importing that constant
~~~
index.ts
import * as fromTrip from './trips.reducer';

export const selectTripState= createFeatureSelector<fromTrip.State>('trips');

export const selectTripEntity = createSelector(
  selectTripState,
  fromTrip.getSelectedTrip
);
~~~

On the other hand if i put all in one file i don't have any error because it use the imported contract from the top of the file
~~~
//trips.reducer.d.ts
export const selectTripEntity = import("@ngrx/store").MemoizedSelector<object,TripContract>;
~~~

#### Expected Behaviour

the extra '../' should not be added

#### Version Information

```
ng-packagr:            4.7.0
@angular/compiler:     7.2.2
rollup:                0.67.4
tsickle:               0.34.0
typescript:            3.2.4
```
@ngrx 7.1.0

**Comment by alan-agius4** [comment]
Does this happen as well if you run `ngc` directly?


It would be great if you would share a minimal reproduction of this issue. Such as a minimal project on github.

**Comment by rams23** [comment]
I have the same problem and I think that it is related is some way to the "import * as " syntax.
Also using ngrx in a multy-entrypoint lib "@mylib".

```
// mylib/core/src/public_api.ts

export interface RequestStatus{...}

// mylib/profile/src/public_api.ts
export * from './reducers/index' //modules can't be found if i remove /index

// mylib/profile/src/reducers/index.ts
import * as fromLocations from './locations.reducer';

export interface ProfileState {
    locations: fromLocations,State;
}

export const getDeleteLocationState = createSelector(
  selectProfile,
  state => state.locations.requestStatus
);

// mylib/profile/src/reducers/locations.reducer.ts
import {RequestStatus} from '@mylib/core'
...
export interface State extends EntityState<Location> {
  locationSelected: string;
  requestStatus: RequestStatus;
}
```

the resulting index.d.ts  contains:
`export declare const getDeleteLocationState: import("@ngrx/store").MemoizedSelector<object, import("../../../../../../../dist/mylib/core").RequestStatus>;`

If I manually  modify the index.d.ts replacing the import path with "@mylib/core" it works fine, also if i remove a '../.
The compilation results of other files where I do not use "import * as " it correctly imports direcly from the entry point name "@mylib/core"

**Comment by morganpizzini** [comment]
Correct,
If I move away declaration from index.ts to the specific file I notice that it works correctly, obviously syntax breaks NgRx architectural pattern, if I can do that by keeping the pattern would be great.
I'll provide soma GitHub repo sample as soon as possible

**Comment by rams23** [comment]
If explicitly import the type and I use  in the selector 'as' like this:
`import {RequestStatus} from '@mylib/core'

export const getDeleteLocationState = createSelector(
  selectProfile,
  state => state.locations.requestStatus as RequestStatus
);` 
it works fine but it is another workaround

**Comment by evenicoulddoit** [comment]
I'm getting the same error as those reporting here, *again* with NgRx selectors, however if I'm reading the situation correctly this issue isn't limited to secondary endpoints.

My build is failing *on a secondary endpoint* (`mypackage/testing`), but the error reported makes it clear that the issue is really with the definition file itself.

Simplified situation:

* `@x/core` package
* `@x/sub` package - has a dependency on `@x/core`
* `@x/sub/selectors.d.ts` produces *relative* and *incorrect* imports. They're one level too deep (e.g. `../../../../dist/core/lib/xyz`)

So, it appears that the definition files are the problem. As @rams23 has commented, updating all of the selector typings to explicitly declare their types is one solution, but both super ugly to do in mass, and also more difficult when using a feature adapter's `getSelectors()` method.

Unfortunately I lack the specific package knowledge to tell if the issue lies:

* In NgRx's selector definitions (given that all the issues to date have cited them)
* An issue with ng-packagr
* An issue with one of the multitude other libraries involve in a `ng build` call

**Comment by amariq** [comment]
Same here.

Building is going fine, but I get output like this:
`export declare const SomeVar: import("../../../dist/somesubmodule/public_api").SomeClass;`
with invalid import path.

Why there is "dist" added to path? It works fine in lib project (as dist folder is present there), but when I import my library into another project (with npm install *.tgz) it fails building because of invalid import paths inside my library.
Basically, correct path for me should be like that:
`export declare const SomeVar: import("../../somesubmodule/public_api").SomeClass;`
_(removed one "../" and "/dist" parts)_

I can't figure out how to fix this. I use minimal ng-package config and trying to stick to default behavior...

**Comment by ac566** [comment]
> If I explicitly import the type and I use in the selector 'as' like this:
> 
> ```
> import {RequestStatus} from '@mylib/core'
> 
> export const getDeleteLocationState = createSelector(
>   selectProfile,
>   state => state.locations.requestStatus as RequestStatus
> );
> ```
> 
> it works fine but it is another workaround

This worked for me as well. In my issue I was using a method from a nother file, that returned a type I hadn't imported. Here is a sample:

```ts
/// file 1
export interface Foo { 
  a: string;
}

/// File 2
import { Foo } from './file1';
export function getMeAFoo(): Foo {
  return { a: 'hello' };
}

/// File 3 
import { Foo } from './file1';
export interface Bar {
  myFoos: {
   [key: string]: Foo
  }
}

/// File 4 (in a separate library)
import { getMeAFoo, Bar } from '@lib/foo';

export class FooBar implements Bar {
  myFoos = {
    foo1: getMeAFoo(),
    foo2: getMeAFoo(),
  } 
}
```

In my case `Foo` has 4 generic parameters, and the `myFoos`, expects that those generic parameters are not all the same.

The fix is to just change the FooBar class to 
```ts
import { getMeAFoo, Bar, Foo } from '@lib/foo';
```

**Comment by leonardochaia** [comment]
I think it's quite clear that if you have a an entrypoint which depends on another and this first entrypoint has a function which wraps a function from the second entrypoint returns it's result (i.e facade) without importing the resulting symbol, definitions generations will be wrong.


```typescript
// /projects/foolib/core/src/lib/models.ts
export interface FooCoreData {
    n: number;
    what: 'ever';
}
```

```typescript
// /projects/foo/core/core.service.ts
import { Injectable } from '@angular/core';
import { FooCoreModule } from './foo-core.module';
import { FooCoreData } from './models';

@Injectable({
    providedIn: FooCoreModule
})
export class FooCoreService {
    public generateData() {
        const data: FooCoreData = {
            n: 123,
            what: 'ever'
        };
        return data;
    }
}
```

```typescript
// projects/foolib/src/lib/foolib.service.ts

import { Injectable } from '@angular/core';
import { FooCoreService } from 'foolib/core';

@Injectable({
  providedIn: 'root'
})
export class FoolibService {
  constructor(private readonly core: FooCoreService) { }
  public wrap() {
    return this.core.generateData();
  }
}

```

Output
```typescript
// /dist/foolib/lib/foolib.service.d.ts
import { FooCoreService } from 'foolib/core';
export declare class FoolibService {
    private readonly core;
    constructor(core: FooCoreService);
    facade(): import("../../../../dist/foolib/core/lib/models").FooCoreData;
}

```

Am I doing something wrong? :/

**Comment by alan-agius4** [comment]
Thanks for reporting this issue. This issue is now obsolete due to changes in the recent releases.
